kotsadm/pkg/store/s3pg/version_store.go

```diff
@@ -267,7 +272,114 @@ func (s S3PGStore) GetAppVersionArchive(appID string, sequence int64) (string, e
+	}
+	defer tx.Rollback()
+
+	newSequence, err := s.createAppVersion(tx, appID, currentSequence, appName, appIcon, kotsKinds)
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to create app version")
+	}
+
+	if err := s.CreateAppVersionArchive(appID, int64(newSequence), filesInDir); err != nil {
+		return int64(0), errors.Wrap(err, "failed to create app version archive")
+	}
+
+	previousArchiveDir := ""
+	if currentSequence != nil {
+		// Get the previous archive, we need this to calculate the diff
+		previousDir, err := s.GetAppVersionArchive(appID, *currentSequence)
+		if err != nil {
+			return int64(0), errors.Wrap(err, "failed to get previous archive")
+		}
+
+		previousArchiveDir = previousDir
+	}
+
+	downstreams, err := s.ListDownstreamsForApp(appID)
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to list downstreams")
+	}
+
+	if err := secrets.ReplaceSecretsInPath(filesInDir); err != nil {
+		return int64(0), errors.Wrap(err, "failed to replace secrets")
+	}
+
+	for _, d := range downstreams {
+		downstreamStatus := "pending"
+		if currentSequence == nil && kotsKinds.Config != nil {
+			downstreamStatus = "pending_config"
+		} else if kotsKinds.Preflight != nil {
+			downstreamStatus = "pending_preflight"
+		}
+
+		if currentSequence != nil {
+			// there's a small chance this is not optimal, but no current code path
+			// will support multiple downstreams, so this is cleaner here for now
+			licenseSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "License")
+			if err != nil {
+				return int64(0), errors.Wrap(err, "failed to marshal license spec")
+			}
+			configSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "Config")
+			if err != nil {
+				return int64(0), errors.Wrap(err, "failed to marshal config spec")
+			}
+			configValuesSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "ConfigValues")
+			if err != nil {
+				return int64(0), errors.Wrap(err, "failed to marshal configvalues spec")
+			}
+
+			// check if version needs additional configuration
+			t, err := kotsconfig.NeedsConfiguration(configSpec, configValuesSpec, licenseSpec)
+			if err != nil {
+				return int64(0), errors.Wrap(err, "failed to check if version needs configuration")
+			}
+			if t {
+				downstreamStatus = "pending_config"
+			}
+		}
+
+		diffSummary, diffSummaryError := "", ""
+		if currentSequence != nil {
+			// diff this release from the last release
+			diff, err := downstream.DiffAppVersionsForDownstream(d.Name, filesInDir, previousArchiveDir, kotsKinds.KustomizeVersion())
+			if err != nil {
+				diffSummaryError = errors.Wrap(err, "failed to diff").Error()
+			} else {
+				b, err := json.Marshal(diff)
+				if err != nil {
+					diffSummaryError = errors.Wrap(err, "failed to marshal diff").Error()
+				}
+				diffSummary = string(b)
+			}
+		}
+
+		commitURL, err := gitops.CreateGitOpsDownstreamCommit(appID, d.ClusterID, int(newSequence), filesInDir, d.Name)
```

Comment:
this is wrong, the store shouldn't call gitops.

Reply:
I think the correct way to solve this is to create gitops commit asynchronously.  Similar to how operator applies downstreams.